src/protocols/secure_channel/CASESession.cpp

```diff
@@ -646,6 +670,79 @@ CHIP_ERROR CASESession::RecoverInitiatorIpk()
     return CHIP_NO_ERROR;
 }
 
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+void CASESession::HandleConnectionComplete(Transport::ActiveConnectionState * conn, CHIP_ERROR err)
+{
+    char peerAddrBuf[chip::Transport::PeerAddress::kMaxToStringSize];
+    VerifyOrReturn(conn != nullptr);
+
+    Transport::PeerAddress peerAddr = conn->mPeerAddr;
+    peerAddr.ToString(peerAddrBuf);
+    CASESession * caseSession = reinterpret_cast<CASESession *>(conn->mAppState->appContext);
+    VerifyOrReturn(caseSession != nullptr);
+
+    // Exit and disconnect if connection setup encountered an error.
+    SuccessOrExit(err);
+
+    ChipLogDetail(SecureChannel, "TCP Connection established with %s before session establishment", peerAddrBuf);
+
+    // Associate the connection with the current unauthenticated session for the
+    // CASE exchange.
+    caseSession->mExchangeCtxt->GetSessionHandle()->AsUnauthenticatedSession()->SetTCPConnection(conn);
+
+    // Associate the connection with the current secure session that is being
+    // set up.
+    caseSession->mSecureSessionHolder.Get().Value()->AsSecureSession()->SetTCPConnection(conn);
+
+    // Retain the initiator exchange until CASE interaction is complete.
+    // The Retain() is necessary to prevent the automatic release of the ExchangeContext
+    // after sending the last message of the Exchange and before the call to
+    // Finish(). This happens because the last sent message does not require an
+    // Ack back on the Exchange and so the ExchangeContext is released after
+    // sending the last message, e.g., a StatusReport.
+    // The corresponding Release() is in PairingSession::Finish()
+    caseSession->mExchangeCtxt->Retain();
+
+    // Send Sigma1 after connection is established for sessions over TCP
+    err = caseSession->SendSigma1();
+    SuccessOrExit(err);
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(SecureChannel, "Connection establishment failed: %s", ErrorStr(err));
+
+        // Release the retained ExchangeContext above if an error happened.
+        caseSession->mExchangeCtxt->Release();
```

Comment:
We can get here with err an error _without having made the retain call_.  

Can we please write this code in a less error-prone way?  What happened with my suggestion that we use an ExchangeHandle to hold the exchange, so we are holding a reference to it, period?